File: reference/src/policy/rego-evaluator.ts
/**
 * Rego-like expression evaluator.
 * Implements a subset of OPA Rego syntax for policy evaluation.
 * Supports property access (dot/bracket), comparisons, functions, and negation.
 * <!-- Source: PRD Section 10.3 -->
 */

import type { ExpressionEvaluator } from './expression.js';

/**
 * Resolve a property path that may include bracket notation.
 * E.g., `input.labels["env"]` or `input.items[0].name`
 */
function resolveRegoPath(path: string, context: Record<string, unknown>): unknown {
  // Tokenize: split on dots but handle bracket access
  const tokens: string[] = [];
  let current = '';
  for (let i = 0; i < path.length; i++) {
    const ch = path[i];
    if (ch === '.') {
      if (current) tokens.push(current);
      current = '';
    } else if (ch === '[') {
      if (current) tokens.push(current);
      current = '';
      const end = path.indexOf(']', i);
      if (end === -1) return undefined;
      let key = path.slice(i + 1, end);
      // Strip quotes from string keys
      if (
        (key.startsWith('"') && key.endsWith('"')) ||
        (key.startsWith("'") && key.endsWith("'"))
      ) {
        key = key.slice(1, -1);
      }
      tokens.push(key);
      i = end;
    } else {
      current += ch;
    }
  }
  if (current) tokens.push(current);

  let value: unknown = context;
  for (const token of tokens) {
    if (value === null || value === undefined) return undefined;
    if (Array.isArray(value)) {
      const idx = parseInt(token, 10);
      if (!isNaN(idx)) {
        value = value[idx];
        continue;
      }
    }
    if (typeof value === 'object') {
      value = (value as Record<string, unknown>)[token];
    } else {
      return undefined;
    }
  }
  return value;
}

function parseRegoValue(token: string, context: Record<string, unknown>): unknown {
  const trimmed = token.trim();
  // String literals
  if (
    (trimmed.startsWith('"') && trimmed.endsWith('"')) ||
    (trimmed.startsWith("'") && trimmed.endsWith("'"))
  ) {
    return trimmed.slice(1, -1);
  }
  // Boolean
  if (trimmed === 'true') return true;
  if (trimmed === 'false') return false;
  // Null
  if (trimmed === 'null') return null;
  // Number
  const num = Number(trimmed);
  if (!isNaN(num) && trimmed !== '') return num;
  // Property path
  return resolveRegoPath(trimmed, context);
}

/** Built-in Rego functions. */
function evaluateFunction(name: string, args: string[], context: Record<string, unknown>): unknown {
  switch (name) {
    case 'count': {
      const val = parseRegoValue(args[0], context);
      if (Array.isArray(val)) return val.length;
      if (typeof val === 'string') return val.length;
      if (val && typeof val === 'object') return Object.keys(val).length;
      return 0;
    }
    case 'startswith': {
      const str = parseRegoValue(args[0], context);
      const prefix = parseRegoValue(args[1], context);
      return typeof str === 'string' && typeof prefix === 'string' && str.startsWith(prefix);
    }
    case 'endswith': {
      const str = parseRegoValue(args[0], context);
      const suffix = parseRegoValue(args[1], context);
      return typeof str === 'string' && typeof suffix === 'string' && str.endsWith(suffix);
    }
    case 'contains': {
      const str = parseRegoValue(args[0], context);
      const sub = parseRegoValue(args[1], context);
      if (typeof str === 'string' && typeof sub === 'string') return str.includes(sub);
      if (Array.isArray(str)) return str.includes(sub);
      return false;
    }
    case 'trim': {
      const str = parseRegoValue(args[0], context);
      return typeof str === 'string' ? str.trim() : str;
    }
    case 'lower': {
      const str = parseRegoValue(args[0], context);
      return typeof str === 'string' ? str.toLowerCase() : str;
    }
    case 'upper': {
      const str = parseRegoValue(args[0], context);
      return typeof str === 'string' ? str.toUpperCase() : str;
    }
    default:
      return undefined;
  }
}

/** Extract function call: `funcName(arg1, arg2)` */
function tryParseFunction(expr: string): { name: string; args: string[] } | null {
  const match = /^([a-z_]+)\((.+)\)$/s.exec(expr.trim());
  if (!match) return null;
  // Split args on top-level commas (not inside quotes or parens)
  const argsStr = match[2];
  const args: string[] = [];
  let depth = 0;
  let current = '';
  for (const ch of argsStr) {
    if (ch === '(' || ch === '[') depth++;
    else if (ch === ')' || ch === ']') depth--;
    else if (ch === ',' && depth === 0) {
      args.push(current.trim());
      current = '';
      continue;
    }
    current += ch;
  }
  if (current.trim()) args.push(current.trim());
  return { name: match[1], args };
}

function evaluateRegoAtomic(expression: string, context: Record<string, unknown>): boolean {
  const trimmed = expression.trim();

  // Negation: `not <expr>`
  if (trimmed.startsWith('not ')) {
    return !evaluateRegoAtomic(trimmed.slice(4), context);
  }

  // `some x in collection` quantifier (simplified: checks non-empty)
  const someMatch = /^some\s+\w+\s+in\s+(.+)$/.exec(trimmed);
  if (someMatch) {
    const val = parseRegoValue(someMatch[1], context);
    if (Array.isArray(val)) return val.length > 0;
    return val !== null && val !== undefined;
  }

  // Comparison operators
  for (const op of ['>=', '<=', '!=', '==', '>', '<']) {
    const opIdx = trimmed.indexOf(` ${op} `);
    if (opIdx !== -1) {
      const leftExpr = trimmed.slice(0, opIdx).trim();
      const rightExpr = trimmed.slice(opIdx + op.length + 2).trim();

      // Check if left side is a function call
      let leftVal: unknown;
      const fnCall = tryParseFunction(leftExpr);
      if (fnCall) {
        leftVal = evaluateFunction(fnCall.name, fnCall.args, context);
      } else {
        leftVal = parseRegoValue(leftExpr, context);
      }

      let rightVal: unknown;
      const rightFn = tryParseFunction(rightExpr);
      if (rightFn) {
        rightVal = evaluateFunction(rightFn.name, rightFn.args, context);
      } else {
        rightVal = parseRegoValue(rightExpr, context);
      }

      const lNum = typeof leftVal === 'number' ? leftVal : Number(leftVal);
      const rNum = typeof rightVal === 'number' ? rightVal : Number(rightVal);
      const numericOk = !isNaN(lNum) && !isNaN(rNum);

      switch (op) {
        case '>=':
          return numericOk && lNum >= rNum;
        case '<=':
          return numericOk && lNum <= rNum;
        case '>':
          return numericOk && lNum > rNum;
        case '<':
          return numericOk && lNum < rNum;
        case '==':
          if (numericOk) return lNum === rNum;
          return leftVal === rightVal;
        case '!=':
          if (numericOk) return lNum !== rNum;
          return leftVal !== rightVal;
      }
    }
  }

  // Boolean function call: `startswith(name, "ai-")`
  const fnCall = tryParseFunction(trimmed);
  if (fnCall) {
    const result = evaluateFunction(fnCall.name, fnCall.args, context);
    return Boolean(result);
  }

  // Truthiness
  const val = parseRegoValue(trimmed, context);
  return Boolean(val);
}

/**
 * Create a Rego-like expression evaluator.
 *
 * Supported syntax:
 * - Property access: `input.metadata.name`, `input.labels["env"]`
 * - Comparisons: `==`, `!=`, `>=`, `<=`, `>`, `<`
 * - Functions: `count()`, `startswith()`, `endswith()`, `contains()`, `lower()`, `upper()`, `trim()`
 * - Negation: `not <expr>`
 * - Quantifier: `some x in collection` (checks non-empty)
 * - Logical: `;` (AND)
 */
export function createRegoEvaluator(): ExpressionEvaluator {
  return {
    evaluate(expression: string, context: Record<string, unknown>): boolean {
      // Rego uses `;` for AND (rule body conjunction)
      if (expression.includes(';')) {
        const parts = expression.split(';');
        return parts.every((p) => evaluateRegoAtomic(p.trim(), context));
      }
      return evaluateRegoAtomic(expression, context);
    },

    validate(expression: string): { valid: boolean; error?: string } {
      if (!expression || expression.trim().length === 0) {
        return { valid: false, error: 'Expression is empty' };
      }
      return { valid: true };
    },
  };
}

File: sdk-python/src/ai_sdlc/policy/rego_evaluator.py
"""Rego-like expression evaluator.

Implements a subset of OPA Rego syntax for policy evaluation.
Supports property access (dot/bracket), comparisons, functions, and negation.
"""

from __future__ import annotations

import re
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ai_sdlc.policy.expression import ExpressionEvaluator


def _resolve_rego_path(path: str, context: dict[str, Any]) -> Any:
    """Resolve a property path that may include bracket notation.

    E.g., ``input.labels["env"]`` or ``input.items[0].name``
    """
    tokens: list[str] = []
    current = ""
    i = 0
    while i < len(path):
        ch = path[i]
        if ch == ".":
            if current:
                tokens.append(current)
            current = ""
        elif ch == "[":
            if current:
                tokens.append(current)
            current = ""
            end = path.find("]", i)
            if end == -1:
                return None
            key = path[i + 1 : end]
            # Strip quotes from string keys
            if (key.startswith('"') and key.endswith('"')) or (
                key.startswith("'") and key.endswith("'")
            ):
                key = key[1:-1]
            tokens.append(key)
            i = end
        else:
            current += ch
        i += 1
    if current:
        tokens.append(current)

    value: Any = context
    for token in tokens:
        if value is None:
            return None
        if isinstance(value, list):
            try:
                idx = int(token)
                value = value[idx]
                continue
            except (ValueError, IndexError):
                pass
        if isinstance(value, dict):
            value = value.get(token)
        else:
            return None
    return value


def _parse_rego_value(token: str, context: dict[str, Any]) -> Any:
    trimmed = token.strip()
    # String literals
    if (trimmed.startswith('"') and trimmed.endswith('"')) or (
        trimmed.startswith("'") and trimmed.endswith("'")
    ):
        return trimmed[1:-1]
    if trimmed == "true":
        return True
    if trimmed == "false":
        return False
    if trimmed == "null":
        return None
    # Number
    try:
        if "." in trimmed:
            return float(trimmed)
        return int(trimmed)
    except ValueError:
        pass
    # Property path
    return _resolve_rego_path(trimmed, context)


def _evaluate_function(
    name: str, args: list[str], context: dict[str, Any]
) -> Any:
    """Built-in Rego functions."""
    if name == "count":
        val = _parse_rego_value(args[0], context)
        if isinstance(val, (list, str)):
            return len(val)
        if isinstance(val, dict):
            return len(val)
        return 0
    if name == "startswith":
        s = _parse_rego_value(args[0], context)
        prefix = _parse_rego_value(args[1], context)
        return isinstance(s, str) and isinstance(prefix, str) and s.startswith(prefix)
    if name == "endswith":
        s = _parse_rego_value(args[0], context)
        suffix = _parse_rego_value(args[1], context)
        return isinstance(s, str) and isinstance(suffix, str) and s.endswith(suffix)
    if name == "contains":
        s = _parse_rego_value(args[0], context)
        sub = _parse_rego_value(args[1], context)
        if isinstance(s, str) and isinstance(sub, str):
            return sub in s
        if isinstance(s, list):
            return sub in s
        return False
    if name == "trim":
        s = _parse_rego_value(args[0], context)
        return s.strip() if isinstance(s, str) else s
    if name == "lower":
        s = _parse_rego_value(args[0], context)
        return s.lower() if isinstance(s, str) else s
    if name == "upper":
        s = _parse_rego_value(args[0], context)
        return s.upper() if isinstance(s, str) else s
    return None


def _try_parse_function(expr: str) -> tuple[str, list[str]] | None:
    """Extract function call: ``funcName(arg1, arg2)``."""
    m = re.match(r"^([a-z_]+)\((.+)\)$", expr.strip(), re.DOTALL)
    if not m:
        return None
    args_str = m.group(2)
    args: list[str] = []
    depth = 0
    current = ""
    for ch in args_str:
        if ch in ("(", "["):
            depth += 1
        elif ch in (")", "]"):
            depth -= 1
        elif ch == "," and depth == 0:
            args.append(current.strip())
            current = ""
            continue
        current += ch
    if current.strip():
        args.append(current.strip())
    return (m.group(1), args)


def _to_number(val: Any) -> float | None:
    if isinstance(val, (int, float)):
        return float(val)
    try:
        return float(val)
    except (TypeError, ValueError):
        return None


def _evaluate_rego_atomic(expression: str, context: dict[str, Any]) -> bool:
    trimmed = expression.strip()

    # Negation: ``not <expr>``
    if trimmed.startswith("not "):
        return not _evaluate_rego_atomic(trimmed[4:], context)

    # ``some x in collection`` quantifier (simplified: checks non-empty)
    some_match = re.match(r"^some\s+\w+\s+in\s+(.+)$", trimmed)
    if some_match:
        val = _parse_rego_value(some_match.group(1), context)
        if isinstance(val, list):
            return len(val) > 0
        return val is not None

    # Comparison operators
    for op in (">=", "<=", "!=", "==", ">", "<"):
        spaced = f" {op} "
        op_idx = trimmed.find(spaced)
        if op_idx != -1:
            left_expr = trimmed[:op_idx].strip()
            right_expr = trimmed[op_idx + len(spaced) :].strip()

            # Check if left side is a function call
            fn_call = _try_parse_function(left_expr)
            left_val = (
                _evaluate_function(fn_call[0], fn_call[1], context)
                if fn_call
                else _parse_rego_value(left_expr, context)
            )

            right_fn = _try_parse_function(right_expr)
            right_val = (
                _evaluate_function(right_fn[0], right_fn[1], context)
                if right_fn
                else _parse_rego_value(right_expr, context)
            )

            l_num = _to_number(left_val)
            r_num = _to_number(right_val)
            numeric_ok = l_num is not None and r_num is not None

            if op == ">=":
                return numeric_ok and l_num >= r_num  # type: ignore[operator]
            if op == "<=":
                return numeric_ok and l_num <= r_num  # type: ignore[operator]
            if op == ">":
                return numeric_ok and l_num > r_num  # type: ignore[operator]
            if op == "<":
                return numeric_ok and l_num < r_num  # type: ignore[operator]
            if op == "==":
                if numeric_ok:
                    return l_num == r_num
                return bool(left_val == right_val)
            if op == "!=":
                if numeric_ok:
                    return l_num != r_num
                return bool(left_val != right_val)

    # Boolean function call
    fn_call = _try_parse_function(trimmed)
    if fn_call:
        result = _evaluate_function(fn_call[0], fn_call[1], context)
        return bool(result)

    # Truthiness
    val = _parse_rego_value(trimmed, context)
    return bool(val)


class _RegoEvaluator:
    def evaluate(self, expression: str, context: dict[str, Any]) -> bool:
        # Rego uses ``;`` for AND (rule body conjunction)
        if ";" in expression:
            parts = expression.split(";")
            return all(_evaluate_rego_atomic(p.strip(), context) for p in parts)
        return _evaluate_rego_atomic(expression, context)

    def validate(self, expression: str) -> dict[str, Any]:
        if not expression or expression.strip() == "":
            return {"valid": False, "error": "Expression is empty"}
        return {"valid": True}


def create_rego_evaluator() -> ExpressionEvaluator:
    """Create a Rego-like expression evaluator.

    Supported syntax:
    - Property access: ``input.metadata.name``, ``input.labels["env"]``
    - Comparisons: ``==``, ``!=``, ``>=``, ``<=``, ``>``, ``<``
    - Functions: ``count()``, ``startswith()``, ``endswith()``, ``contains()``,
      ``lower()``, ``upper()``, ``trim()``
    - Negation: ``not <expr>``
    - Quantifier: ``some x in collection`` (checks non-empty)
    - Logical: ``;`` (AND)
    """
    return _RegoEvaluator()
